// react-image-uploader/lib/DropZone.js
'use strict';

exports.__esModule = true;

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var DropZone = (function (_React$Component) {
  _inherits(DropZone, _React$Component);

  function DropZone() {
    _classCallCheck(this, DropZone);

    _React$Component.apply(this, arguments);

    this.state = {
      onDragOver: false
    };
  }

  DropZone.prototype.render = function render() {
    var _props = this.props;
    var alt = _props.alt;
    var image = _props.image;
    var onUploadPrompt = _props.onUploadPrompt;
    var onRender = _props.onRender;
    var onDragOver = this.state.onDragOver;

    // allow the user to render their own elements
    if (onRender) {
      var el = onRender(this.props, this.state, this);
      if (image) return el;
      return _react2['default'].cloneElement(el, {
        onDragEnter: this.onDragEnter.bind(this),
        onDragOver: this.onDragOver.bind(this),
        onDragLeave: this.onDragLeave.bind(this),
        onDrop: this.props.onDrop
      });
    }

    if (image) {
      var _styles = {
        width: '230px',
        height: '250px',
        backgroundSize: 'cover',
        backgroundImage: 'url(' + image + ')'
      };

      return _react2['default'].createElement('div', {
        role: 'image',
        style: _styles,
        'aria-label': alt
      });
    }

    var styles = {
      width: '230px',
      height: '250px',
      border: '1px solid #ddd'
    };

    if (onDragOver) {
      styles.borderColor = '#08c';
    }

    return _react2['default'].createElement(
      'div',
      {
        onDragEnter: this.onDragEnter.bind(this),
        onDragOver: this.onDragOver.bind(this),
        onDragLeave: this.onDragLeave.bind(this),
        onDrop: this.onDrop.bind(this),
        style: styles },
      _react2['default'].createElement(
        'button',
        { onClick: onUploadPrompt },
        'Add Photo'
      )
    );
  };

  DropZone.prototype.onDragEnter = function onDragEnter(e) {
    e.preventDefault();
    this.setState({ onDragOver: true });
  };

  DropZone.prototype.onDragOver = function onDragOver(e) {
    e.preventDefault();
  };

  DropZone.prototype.onDragLeave = function onDragLeave() {
    this.setState({ onDragOver: false });
  };

  DropZone.prototype.onDrop = function onDrop(e) {
    this.setState({ onDragOver: false });
    e.preventDefault();
    this.props.onDrop(e);
  };

  _createClass(DropZone, null, [{
    key: 'propTypes',
    value: {
      image: _react.PropTypes.string,
      error: _react.PropTypes.any,
      alt: _react.PropTypes.string,
      onDrop: _react.PropTypes.func.isRequired,
      onRequestRemove: _react.PropTypes.func,
      onUploadPrompt: _react.PropTypes.func.isRequired
    },
    enumerable: true
  }]);

  return DropZone;
})(_react2['default'].Component);

exports['default'] = DropZone;
module.exports = exports['default'];
